### ISSUE TITLE
Workflow is running even if a ValidatingWebhookConfiguration rejects the pods

### ISSUE THREAD

**Issue by karabill** [argoproj/argo-workflows#5245]
## Summary 

I have a validating webhooks which rejects all pods in specific namespaces. I submit a workflow and the pod is not created because it is rejected by the webhook (as it is expected). However, the workflow is still Running. I would expect the workflow to be in the Error state. I have created a repo with some instructions, if you want to reproduce it in a kind cluster: https://github.com/karabill/argo_reject_operator_problem. Note that this is happening only in argo v2.12.* (I tried v2.12.9 and a couple of other v2.12.*). I was not able to reproduce it in v2.11.8.

## Diagnostics

What Kubernetes provider are you using? 
AWS or kind

What version of Argo Workflows are you running? 
* v2.12.9

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1alpha1","kind":"Workflow","metadata":{"annotations":{},"name":"myworkflow85","namespace":"test-workflow"},"spec":{"entrypoint":"start","templates":[{"container":{"args":["echo 'it should not run'"],"command":["/bin/sh","-c"],"image":"alpine:3.7"},"name":"start"}]}}
  creationTimestamp: "2021-03-01T16:01:50Z"
  generation: 2
  labels:
    workflows.argoproj.io/phase: Running
  name: myworkflow85
  namespace: test-workflow
  resourceVersion: "1568"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/test-workflow/workflows/myworkflow85
  uid: 63c1aa45-501a-4a6a-9f69-ddd865084596
spec:
  arguments: {}
  entrypoint: start
  templates:
  - arguments: {}
    container:
      args:
      - echo 'it should not run'
      command:
      - /bin/sh
      - -c
      image: alpine:3.7
      name: ""
      resources: {}
    inputs: {}
    metadata: {}
    name: start
    outputs: {}
status:
  finishedAt: null
  nodes:
    myworkflow85:
      displayName: myworkflow85
      finishedAt: "2021-03-01T16:01:50Z"
      id: myworkflow85
      message: 'admission webhook "pod.validation.webhook" denied the request: Pod
        is not allowed'
      name: myworkflow85
      phase: Error
      progress: 1/1
      startedAt: "2021-03-01T16:01:50Z"
      templateName: start
      templateScope: local/myworkflow85
      type: Pod
  phase: Running
  progress: 1/1
  startedAt: "2021-03-01T16:01:50Z"
```

```
time="2021-03-01T16:01:50.893Z" level=info msg="Processing workflow" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.893Z" level=info msg="Updated phase  -> Running" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.893Z" level=info msg="Pod node myworkflow85 initialized Pending" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.899Z" level=info msg="Failed to create pod myworkflow85 (myworkflow85): admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.899Z" level=error msg="Mark error node" error="admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow nodeName=myworkflow85 workflow=myworkflow85
time="2021-03-01T16:01:50.899Z" level=info msg="node myworkflow85 phase Pending -> Error" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.899Z" level=info msg="node myworkflow85 message: admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.900Z" level=info msg="node myworkflow85 finished: 2021-03-01 16:01:50.900014662 +0000 UTC" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.900Z" level=error msg="error in entry template execution" error="admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow85
time="2021-03-01T16:01:50.908Z" level=info msg="Workflow update successful" namespace=test-workflow phase=Running resourceVersion=1568 workflow=myworkflow85
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
Can you confirm if this is a regression from v2.11, or that you just could not repro in v2.11?

**Comment by karabill** [comment]
I could not reproduce it in v2.11, i.e. in v2.11 the state of the workflow was always Error. I also tried a couple of other older version (e.g. 2.9) but the workflows were always in Error. I only saw this behaviour in v2.12.

**Comment by alexec** [comment]
I believe the problem is caused by the pod failing as the error is a 403 error, which is considered "transient".

```golang
	node, err := woc.executeTemplate(ctx, woc.wf.ObjectMeta.Name, &wfv1.WorkflowStep{Template: woc.execWf.Spec.Entrypoint}, tmplCtx, woc.execWf.Spec.Arguments, &executeTemplateOpts{})
	if err != nil {
		woc.log.WithError(err).Error("error in entry template execution")
		// we wrap this error up to report a clear message
		x := fmt.Errorf("error in entry template execution: %w", err)
		switch err {
		case ErrDeadlineExceeded:
			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", x.Error())
		case ErrParallelismReached:
		default:
/// NEW LINES
			if !errorsutil.IsTransientErr(err) && !woc.wf.Status.Phase.Completed() && os.Getenv("BUBBLE_ENTRY_TEMPLATE_ERR") != "false" {
				woc.markWorkflowError(ctx, x)
			}
		}
		return
	}
```

**Comment by alexec** [comment]
@karabill can you confirm that `BUBBLE_ENTRY_TEMPLATE_ERR` is not `false`?

**Comment by karabill** [comment]
@alexec I assume that you are talking about the workflow-controller. I do not see it in the pod spec, so I assume that `os.Getenv` will return an empty string. 

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2021-03-01T15:57:55Z"
  generateName: workflow-controller-67d6f85d99-
  labels:
    app: workflow-controller
    pod-template-hash: 67d6f85d99
  name: workflow-controller-67d6f85d99-h7qvq
  namespace: argo
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: workflow-controller-67d6f85d99
    uid: 03dc9d23-0f62-4aa7-b40e-9fa8da39ef26
  resourceVersion: "720"
  selfLink: /api/v1/namespaces/argo/pods/workflow-controller-67d6f85d99-h7qvq
  uid: e39130aa-4c2a-4827-bce5-f3a1e602fc21
spec:
  containers:
  - args:
    - --configmap
    - workflow-controller-configmap
    - --executor-image
    - argoproj/argoexec:v2.12.9
    command:
    - workflow-controller
    image: argoproj/workflow-controller:v2.12.9
    imagePullPolicy: IfNotPresent
    name: workflow-controller
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: argo-token-qqjx7
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: mytest-control-plane
  nodeSelector:
    kubernetes.io/os: linux
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: argo
  serviceAccountName: argo
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: argo-token-qqjx7
    secret:
      defaultMode: 420
      secretName: argo-token-qqjx7
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-03-01T15:57:55Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2021-03-01T15:58:22Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2021-03-01T15:58:22Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2021-03-01T15:57:55Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://d7bc32fbc4204d865f45a5841372b0b4642cb243476cb6d6164c7206d3e5a273
    image: docker.io/argoproj/workflow-controller:v2.12.9
    imageID: docker.io/argoproj/workflow-controller@sha256:e00d3c59837c7b79a5091a4bafbd7a7fd31c23ccfb24e938d1b3b4ecf75b7708
    lastState: {}
    name: workflow-controller
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2021-03-01T15:58:21Z"
  hostIP: 172.18.0.2
  phase: Running
  podIP: 10.244.0.6
  podIPs:
  - ip: 10.244.0.6
  qosClass: BestEffort
  startTime: "2021-03-01T15:57:55Z"
```

**Comment by karabill** [comment]
I just read my description again. I forgot to mention that this error is not deterministic. E.g. if I submit 100 workflows, around 10-20% are still running. The others' state is Error.

**Comment by alexec** [comment]
Failed to reproduce this using K3S. Try again tomorrow.

**Comment by karabill** [comment]
In the description, I have a link to a repository and the instructions to reproduce it using [kind](https://kind.sigs.k8s.io). Maybe it helps. Moreover, I submitted 100 workflows yesterday. After 10 minutes around 10% were still running. Today (after 15 hours) all of them are in Error state.

**Comment by alexec** [comment]
I've created an engineering build with extra logging to help diagnose this issue. Can I please ask you to test it out? `argoproj/argoexec:dev-debug`.

**Comment by karabill** [comment]
Thank you. I used it. However, I was not able to reproduce the problem. All workflows' state was error (I tried it a couple of times).

What is the argo's version upon which you based your dev-debug? I am asking this because I had to add the `LEADER_ELECTION_IDENTITY`ENV  in https://github.com/karabill/argo_reject_operator_problem/blob/main/yamls/install-argo.yaml#L426. Is the dev-debug based on the `v2.12.9` or it is based on another (newer) version?

Just in case you need them, the logs that I go are:
```
time="2021-03-03T09:38:29.765Z" level=info msg="Processing workflow" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.768Z" level=info msg="Updated phase  -> Running" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.769Z" level=info msg="Pod node myworkflow75 initialized Pending" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=info msg="Failed to create pod myworkflow75 (myworkflow75): admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=error msg="Mark error node" error="admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow nodeName=myworkflow75 workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=info msg="node myworkflow75 phase Pending -> Error" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=info msg="node myworkflow75 message: admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=info msg="node myworkflow75 finished: 2021-03-03 09:38:29.774724126 +0000 UTC" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=error msg="error in entry template execution" error="admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.774Z" level=error bubbleEntryTemplateErr= completed=false namespace=test-workflow transientErr=false workflow=myworkflow75
time="2021-03-03T09:38:29.775Z" level=info msg="Updated phase Running -> Error" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.775Z" level=info msg="Updated message  -> error in entry template execution: admission webhook \"pod.validation.webhook\" denied the request: Pod is not allowed" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.775Z" level=info msg="Marking workflow completed" namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.775Z" level=info msg="Checking daemoned children of " namespace=test-workflow workflow=myworkflow75
time="2021-03-03T09:38:29.792Z" level=info msg="Workflow update successful" namespace=test-workflow phase=Error resourceVersion=4700 workflow=myworkflow75
```

**Comment by alexec** [comment]
Error state is correct state?

**Comment by karabill** [comment]
Yes, the error state is the correct state (i.e. expected state).

**Comment by alexec** [comment]
Ok. So this is fixed in v3.

**Comment by karabill** [comment]
I have not tried v3, because it was in RC. I will try again when it is released. In the meantime, we can close this issue, it is up to you. @alexec thank you for your help.

**Comment by alexec** [comment]
No problem. I'll discus with the team if we plan to fix in v2.12, but I think it is unlikely due to the small number of people impacted.
